Restore client rpc tests

Oscar-Pepper opened the pull request.

Oscar-Pepper committed:
created new test fixtures crate

Oscar-Pepper committed:
started restoring test fixtures

Oscar-Pepper committed:
use latest stable zingolib tag

Oscar-Pepper committed:
part way through fixing build errors

Oscar-Pepper committed:
client rpc test fixtures building successfully

Oscar-Pepper committed:
move chain cache to client_rpc_test_fixtures crate

Oscar-Pepper committed:
restore client rpc tests

Oscar-Pepper committed:
solve merge conlficts with latest dev

Oscar-Pepper committed:
Merge branch 'fix_zainod_config' into restore_client_rpc_tests

Oscar-Pepper committed:
update zcashd chain cache and fix get_block_range_out_of_bounds test

Oscar-Pepper committed:
remove git ignore

Oscar-Pepper committed:
cargo fmt

Oscar-Pepper committed:
update out of bound test assertions

Oscar-Pepper committed:
update fix for zebra heartwood-nu5 validation issues for chain index integration testing

Oscar-Pepper committed:
update chain cache

Oscar-Pepper committed:
solve merge conflicts with latest dev

Oscar-Pepper committed:
fix comment error

nachog00 commented on client_rpc_test_fixtures/src/lib.rs:
```diff
@@ -79,16 +79,17 @@ use zcash_local_net::{
 };
 use zingo_test_vectors::{REG_O_ADDR_FROM_ABANDONART, ZEBRAD_DEFAULT_MINER, seeds};
 
+// temporarily setting nu5+ to height 2 until crates are updated to zebra 3.0 due to issues with testing zainos chain index integration
```
is there some issue on zebra that explains why this is needed? IIUC it's solved already and coming in with zebra 3 right?
I would add as many references on this comment as possible so that this change isn't so cryptic in the future.

Oscar-Pepper replied:
i cant actually find the zebra issue related to this. I think the comment clearly states that after we have updated to zebra 3 we could set this back to all ones but its not critical either way TBH. maybe you could add to this if you think it could be expanded?

nachog00 replied:
I'm not so familiar with the implicancies, is it a we COULD or a we SHOULD change them?
If it's the later, then at least making an issue here on infra for it so it doesn't fly under the radar. otherwise, the comment is fine i think.

Oscar-Pepper replied:
its more we "could", having these activation heights do not change the test but its just makes more sense to use defaults of all 1s if we can. There should only be a deviation from this if there is a reason... which there currently is one due to the zebra bug

nachog00 commented on zcash_local_net/tests/integration.rs:
```diff
@@ -120,26 +125,124 @@ async fn localnet_launch_multiple_zebrads_with_cache() {
+                blossom: Some(1),
+                heartwood: Some(1),
+                canopy: Some(1),
+                nu5: Some(2),
```
this seems to be the same temporary "2" values from the `DEFAULT_ACTIVATION_HEIGHTS`, why not use the variable directly? otherwise, updating the variable later will likely leave these rogue changes around.

Oscar-Pepper replied:
i would prefer to keep these crates not depending on one another for the time being. I actually tried this 1st time around.

Oscar-Pepper replied:
oh it was also related to zcash_local_net already having a default which is depended on by external crates. we decided it was not correct to change this default as it could have larger implications so we decided to have a temp config in zcash_local_net and a seperate DRY in client_rpc_tests

nachog00 replied:
ah ok, so different crates, makes sense. Still, it look like the height are the same, so they are the same for the same reason or unrelated?

Oscar-Pepper replied:
they are the same for the same reason yes. they will be removed shortly

nachog00 commented on zcash_local_net/tests/integration.rs:
```diff
@@ -120,26 +125,124 @@ async fn localnet_launch_multiple_zebrads_with_cache() {
+            network_listen_port: None,
+            rpc_listen_port: None,
+            indexer_listen_port: None,
+            configured_activation_heights: ConfiguredActivationHeights {
```
same as above

nachog00 commented on zcash_local_net/tests/integration.rs:
```diff
@@ -120,26 +125,124 @@ async fn localnet_launch_multiple_zebrads_with_cache() {
+    let p = LocalNet::<Zcashd, Lightwalletd>::launch(LocalNetConfig {
+        validator_config: ZcashdConfig {
+            rpc_listen_port: None,
+            configured_activation_heights: ConfiguredActivationHeights {
```
same as above

nachog00 commented on zcash_local_net/tests/integration.rs:
```diff
@@ -120,26 +125,124 @@ async fn localnet_launch_multiple_zebrads_with_cache() {
+            network_listen_port: None,
+            rpc_listen_port: None,
+            indexer_listen_port: None,
+            configured_activation_heights: ConfiguredActivationHeights {
```
same as above

nachog00 commented on zcash_local_net/src/launch.rs:
```diff
@@ -109,6 +109,7 @@ pub(crate) fn wait(
             }
         }
 
+        // TODO: replace with tokio sleep to prevent risk of blocking the runtime
```
could we get this done before merging?

Oscar-Pepper replied:
i think we should address this is a separate PR

fluidvanadium commented on zcash_local_net/tests/integration.rs:
```diff
@@ -122,44 +125,124 @@ async fn localnet_launch_multiple_zebrads_with_cache() {
+                nu7: None,
+            },
+            miner_address: Some(REG_O_ADDR_FROM_ABANDONART),
+            chain_cache: None,
+        },
+        indexer_config: LightwalletdConfig::default(),
+    })
```
This needs to be dry

Oscar-Pepper replied:
i have DRYed these temp configs and added comments to explain better

fluidvanadium commented on client_rpc_test_fixtures/Cargo.toml:
```diff
@@ -0,0 +1,44 @@
+zebra-chain = { workspace = true }
+
+# Zingo
+zingolib = { git = "https://github.com/zingolabs/zingolib.git", tag = "2.0.7", features = [ "testutils" ] }
```
Is there a way to avert this ? Please lets reconsider. The cost is high.

Oscar-Pepper replied:
this is not trivial to remove and not an issue for an isolated crate. zingolib is a great tool for testing responses to lightclients. the high cost you are referring to I suppose is related to trying to re-integrate these test fixtures into zaino test crates? IMO the issue here is not that zingolib is the wrong tool, its that we have mishandled the publib API. the challenge then becomes "how do we fix zingolib's API" instead of "how do we remove zingolib from the client rpc test fixtures"

fluidvanadium replied:
I just put a lot of effort into detangling zingolib from infrastructure. That had to happen because paradoxical dependencies made updates extremely difficult. What happens now if the interface to infrastructure changes? Zingolib needs updated infrastructure and infrastructure needs updated zingolib. This leaves us pulling our own bootstraps. Circular dependencies between multiple repositories is a problem.

Oscar-Pepper replied:
as i said this is an isolated crate, its in the workspace of infrastructure but it is not zcash_local_net. if there are circular dependencies then the issue is not with using zingolib, its deeper than that. The confusion with circular dependencies is not relevant to this test fixture crate... especially when it is not currently being integrated into zaino test crates.

To be clear, these test fixtures can use *any* version of zingolib that suits it to perform the lightclient tasks and zingolib will *never* depend of these test fixtures.... it makes no sense to have these tests in zinfgolib as they are for testing validator and indexers under development

Oscar-Pepper committed:
move client rpc tests to CI nightly only

nachog00 commented on client_rpc_test_fixtures/src/lib.rs:
```diff
@@ -0,0 +1,3552 @@
+    nu6_1: Some(BlockHeight::from_u32(2)),
+};
+
+// TODO: replace with zingo-netutils version when dep graph is stabilised
```
if this can't be solved before merging, let's make an issue for it.

Oscar-Pepper replied:
can you create an issue so we can solve this in a future PR please?

nachog00 commented on client_rpc_test_fixtures/src/lib.rs:
```diff
@@ -0,0 +1,3552 @@
+            .build(connector),
+    )
+}
+// TODO: replace with zingo-netutils version when dep graph is stabilised
```
same as above

nachog00 replied:
```suggestion
// TODO: replace with zingo-netutils version when dep graph is stabilised
// https://github.com/zingolabs/infrastructure/issues/174
```

nachog00 commented on client_rpc_test_fixtures/src/lib.rs:
```diff
@@ -0,0 +1,3552 @@
+            lwd_tx_summaries[i].outgoing_transparent_coins,
+            zainod_tx_summaries[i].outgoing_transparent_coins
+        );
+        // TODO: fix detailed tx summaries in zingolib
```
does this have an issue? let's link it

Oscar-Pepper replied:
no I think we can actually use the transactions directly from the wallet now as this uses zingo 2. again I think we can seperate into a future PR

Oscar-Pepper replied:
https://github.com/zingolabs/infrastructure/issues/176

nachog00 commented on client_rpc_test_fixtures/src/lib.rs:
```diff
@@ -0,0 +1,3552 @@
+            lwd_tx_summaries[i].outgoing_transparent_coins,
+            zainod_tx_summaries[i].outgoing_transparent_coins
+        );
+        // TODO: fix detailed tx summaries in zingolib
```
same as above

Oscar-Pepper replied:
https://github.com/zingolabs/infrastructure/issues/176

Oscar-Pepper committed:
DRY temp configs

nachog00 commented on client_rpc_test_fixtures/src/lib.rs:
```diff
@@ -0,0 +1,3552 @@
+
+/// GetLatestBlock RPC test
+pub async fn get_latest_block() {
+    let zcashd = Zcashd::launch(ZcashdConfig {
```
these first 20 lines seem to be repeated on every single test, we should dry them.

Oscar-Pepper replied:
sure, these tests are not new though so this should not be reviewed as new code

nachog00 commented on client_rpc_test_fixtures/src/lib.rs:
this file is pretty massive, it's even bugging out my GH review page.
we should probably chop it down into a helpers module and a tests module.

Oscar-Pepper replied:
can we do this in a future PR? this is important to have in the codebase before I have to resolve more conflicts

fluidvanadium reviewed (changes requested):
These two ConfirmedActivationHeights happen to be the same. Do we want them to be able to diverge? Or should we consolidate them?

nachog00 commented on client_rpc_test_fixtures/src/lib.rs:
```diff
@@ -0,0 +1,3552 @@
+};
+use zingo_test_vectors::{REG_O_ADDR_FROM_ABANDONART, ZEBRAD_DEFAULT_MINER, seeds};
+
+// temporarily setting nu5+ to height 2 until crates are updated to zebra 3.0 due to issues with testing zainos chain index integration
+const DEFAULT_ACTIVATION_HEIGHTS: ConfiguredActivationHeights = ConfiguredActivationHeights {
+    before_overwinter: Some(1),
+    overwinter: Some(1),
+    sapling: Some(1),
+    blossom: Some(1),
+    heartwood: Some(1),
+    canopy: Some(1),
+    nu5: Some(2),
+    nu6: Some(2),
+    nu6_1: Some(2),
+    nu7: None,
+};
+const DEFAULT_ACTIVATION_HEIGHTS_ZP: LocalNetwork = LocalNetwork {
+    overwinter: Some(BlockHeight::from_u32(1)),
+    sapling: Some(BlockHeight::from_u32(1)),
+    blossom: Some(BlockHeight::from_u32(1)),
+    heartwood: Some(BlockHeight::from_u32(1)),
+    canopy: Some(BlockHeight::from_u32(1)),
+    nu5: Some(BlockHeight::from_u32(2)),
+    nu6: Some(BlockHeight::from_u32(2)),
+    nu6_1: Some(BlockHeight::from_u32(2)),
+};
```
```suggestion

const DEFAULT_ACTIVATION_HEIGHTS: ConfiguredActivationHeights = ConfiguredActivationHeights {
    before_overwinter: Some(1),
    overwinter: Some(1),
    sapling: Some(1),
    blossom: Some(1),
    heartwood: Some(1),
    canopy: Some(1),
    nu5: Some(1),
    nu6: Some(1),
    nu6_1: Some(1),
    nu7: None,
};
const DEFAULT_ACTIVATION_HEIGHTS_ZP: LocalNetwork = LocalNetwork {
    overwinter: Some(BlockHeight::from_u32(1)),
    sapling: Some(BlockHeight::from_u32(1)),
    blossom: Some(BlockHeight::from_u32(1)),
    heartwood: Some(BlockHeight::from_u32(1)),
    canopy: Some(BlockHeight::from_u32(1)),
    nu5: Some(BlockHeight::from_u32(1)),
    nu6: Some(BlockHeight::from_u32(1)),
    nu6_1: Some(BlockHeight::from_u32(1)),
};
```

nachog00 replied:
We should keep the original defaults if adding the new TEMP_ vars, right?

Oscar-Pepper replied:
i dont understand your suggestion... the default must have nu5+ set to height 2 for the client RPC tests until zebra 3.0 update lands everywhere... is my comment confusing?

Oscar-Pepper replied:
> We should keep the original defaults if adding the new TEMP_ vars, right?

no, these are the defaults for this crate and will be updated... they are not depended on by other crates like zcash_local_net is and they are not public

Oscar-Pepper commented:
> These two ConfirmedActivationHeights happen to be the same. Do we want them to be able to diverge? Or should we consolidate them?

Is this regarding the two temp configs I just added in response to your DRY request? These are temporary and will be removed very soon after we have updated to zebra 3.0 across our codebases... I think this is fine as is but you can DRY this with an additional temporary activation heights to be shared by the temporary validator configs if you see fit. I dont think this is the priority for me right now

Oscar-Pepper committed:
fix incorrect reversal of LWDs block hash response to get_latest_block

Oscar-Pepper committed:
fix mut warning

Oscar-Pepper committed:
improve get_mempool_stream test

Oscar-Pepper committed:
fix trailing whitespace

Oscar-Pepper committed:
remove incorrect reversals of txids in get_mempool_tx test to highlight LWD bug

Oscar-Pepper committed:
add a note to the cargo toml about potential future update issues

fluidvanadium commented:
Moved to https://github.com/zingolabs/client_rpc_test_fixtures

Oscar-Pepper commented:
moved to a new repo https://github.com/zingolabs/client_rpc_test_fixtures
